crates/stark_hash/src/hash.rs

```diff
@@ -65,38 +65,94 @@ impl StarkHash {
-    pub fn as_be_bytes(&self) -> &[u8; 32] {
+    pub const fn as_be_bytes(&self) -> &[u8; 32] {
         &self.0
     }
 
     /// Convenience function which extends [StarkHash::from_be_bytes] to work with slices.
-    pub fn from_be_slice(bytes: &[u8]) -> Result<Self, OverflowError> {
+    pub const fn from_be_slice(bytes: &[u8]) -> Result<Self, OverflowError> {
         if bytes.len() > 32 {
             return Err(OverflowError);
         }
 
         let mut buf = [0u8; 32];
-        buf[32 - bytes.len()..].copy_from_slice(bytes);
+        let mut index = 0;
+
+        loop {
+            if index == bytes.len() {
+                break;
+            }
+
+            buf[32 - bytes.len() + index] = bytes[index];
+            index += 1;
+        }
 
         StarkHash::from_be_bytes(buf)
     }
 
-    /// Creates a [StarkHash] from big-endian bytes.
-    ///
-    /// Returns [OverflowError] if not less than the field modulus.
-    pub fn from_be_bytes(bytes: [u8; 32]) -> Result<Self, OverflowError> {
+    #[cfg(fuzzing)]
+    pub fn from_be_bytes_orig(bytes: [u8; 32]) -> Result<Self, OverflowError> {
         // FieldElement::from_repr[_vartime] does the check in a correct way
         match FieldElement::from_repr_vartime(FieldElementRepr(bytes)) {
             Some(field_element) => Ok(Self(field_element.to_repr().0)),
             None => Err(OverflowError),
         }
     }
 
+    /// Creates a [StarkHash] from big-endian bytes.
+    ///
+    /// Returns [OverflowError] if not less than the field modulus.
+    pub const fn from_be_bytes(bytes: [u8; 32]) -> Result<Self, OverflowError> {
+        // ff uses byteorder BigEndian::read_u64_into which uses copy_nonoverlapping(..) and
+        // u64::to_be(), this is essentially the same, though would like to test conclusively
+        #[rustfmt::skip]
+        let limbs = [
+            u64::from_be_bytes([
+                bytes[24], bytes[25], bytes[26], bytes[27],
+                bytes[28], bytes[29], bytes[30], bytes[31],
+            ]),
+            u64::from_be_bytes([
+                bytes[16], bytes[17], bytes[18], bytes[19],
+                bytes[20], bytes[21], bytes[22], bytes[23],
+            ]),
+            u64::from_be_bytes([
+                bytes[8], bytes[9], bytes[10], bytes[11],
+                bytes[12], bytes[13], bytes[14], bytes[15],
+            ]),
+            u64::from_be_bytes([
+                bytes[0], bytes[1], bytes[2], bytes[3],
+                bytes[4], bytes[5], bytes[6], bytes[7],
+            ]),
+        ];
+
+        // this is from expansion, `const MODULUS_LIMBS: FieldElementRepr = [...];`
+        let modulus = [1u64, 0u64, 0u64, 576460752303423505u64];
+
+        let mut borrow = 0;
+        let mut index = 0;
+
+        loop {
+            if index == 4 {
+                break;
+            }
+            borrow = ff::derive::sbb(limbs[index], modulus[index], borrow).1;
+            index += 1;
+        }
```

Comment:
I would have thought `for 0..4 {}` would be valid in a const context, but appears it isn't. Strange limitation.

Reply:
No traits in const, the for loops desugar to iterator trait usage. for has a special error for it which reminded me of this :)